### ISSUE TITLE
Suggestion for future CodeMaid

### ISSUE THREAD

**Issue by heku** [codecadwallader/codemaid#296]
Hi Steve, 
This is just a suggestion so I don't follow your issue format. I've been using CodeMaid for server years, it is a awesome tool, thank you for your great job. This idea in my mind for a long time that could CodeMaid's modules be turn on/off-able such as Microsoft's Productivity Power Tools, it would be more better if each module can be turn on/off in CodeMaid's configuration page. If one module turned off, it will disappear from menus, and it will come back if it turned on. I don't know whether it is possible to achieve this, anyway, thank you in advance.

**Comment by codecadwallader** [comment]
Thank you for the suggestion, and the very nice comments. :)

There was a feature request in our old Trello backlog somewhat similar to this request but it never gained a lot of popularity: https://trello.com/c/O5ldkODW

Let's keep this issue open for awhile and see how many people would benefit from the capability of enabling/disabling features of CodeMaid.  The main tangible effect as you mentioned would be the visibility of menu items, the actual code would really still be the same under the hood.

**Comment by heku** [comment]
Thank you for your consideration, I'm signing up to vote some great features on Trello.

**Comment by codecadwallader** [comment]
Trello is our old backlog - GitHub is now where we track all feature
requests since they added the ability to +1 comments/issues

**Comment by heku** [comment]
@codecadwallader Hi, do you mind I add this feature? I'm going to pull a request for this. My solution is add a checkbox named 'Feature enabled' on each feature's general page, if a feature is disabled, its menu will be hidden. Please let me know if you have any better solution.

**Comment by codecadwallader** [comment]
@heku thank you for asking.  I am open to the feature.  A couple thoughts:

1.  It may be hard to find a 1:1 mapping for every menu item to a configuration page (e.g. sort lines does not have a configuration page).  Another idea is to have a separate page in CodeMaid Options solely dedicated to toggling the visibility of each menu item.  (e.g. General->Menus).

2.  Instead of calling it "Feature enabled" I would prefer something like "menu visibility".  I think the first could be confused as toggling the switch actually has an underlying impact to the code base whereas the latter is clearer that it is just hiding a menu item.

3.  As there may be more menu items over time, and as another pull request is in progress for internationalization, it would be helpful if the list of menu items and their text was dynamically pulled at runtime vs. a hard-coded list in the UI that would have to be manually updated and/or translated.

**Comment by heku** [comment]
@codecadwallader Great suggestions for me, which branch should I start from? master? or ...?

**Comment by codecadwallader** [comment]
Yep, master would be best.  Good luck and feel free to ask questions!

**Comment by heku** [comment]
@codecadwallader , did you ever think of separating features into a few small extensions, i.e. CodeMaid-Cleanup, CodeMaid-FindInSolutionExplorer, CodeMaid-Sapde etc. As CodeMaid's feature is growing, and not all of them needed for every user, so it is much better if user can choose/download those they need only. Separating features is not so difficult (I think), but I don't have any idea about aggregate their options pages (for other VS menus, it's easy to aggregate them by sharing same parent ids), if the options pages were integrated into VS's options (Tools->Options), seems we can aggregate them by sharing same category name, but for current implementation, they are merged in your custom WPF window, I don't know how to archive this, do you have any idea on this?

**Comment by codecadwallader** [comment]
Separating into multiple extensions has been considered in the past, and there may even be an issue laying around somewhere about it.  Since CodeMaid's functionality is largely idle unless invoked, there really isn't any processing advantage to users.. and there's a heavy maintenance overhead to having to publish N number of extensions, have users install N number of extensions, etc.

A long time ago our configuration was inside VS's options (Tools->Options).  For many years it was not a resizable dialog, so we eventually went to having our own configuration window so we could have more flexibility.  It also allows us to have more complex concepts, such as user level settings and solution level settings, that would seem incongruous with VS's options.

**Comment by heku** [comment]
Thanks for the background, make sense to me.

**Comment by Ordisoftware** [comment]
@codecadwallader @heku Need that too to be able to choose what menu item enable or hide like with PowerCommands because I only use the very great Code digging / Spade panel and the Collapse context menu in solution explorer and the Sort in text editor. In fact to do this in a simple manner it may be only one option to shos or hide Codemaid items in the solution explorer context menu.

:bulb: *Another easy way may be to group Codemaid items in the solution explorer root context menu in a single menu as in the Text editor.*

:memo: I would certainly create issues in the future for these suggestions if not exists but here they are quickly:
- [ ] Text Editor : Cleanup selected text.
- [ ] Text Editor : Remove blank lines in selection (no rules, just remove all blank lines).
- [ ] Spade : Rename (open the refactor renaming VS box).

**Comment by Ordisoftware** [comment]
@codecadwallader @heku Need that too to be able to choose what menu item enable or hide like with PowerCommands because I only use the very great Code digging / Spade panel and the Collapse context menu in solution explorer and the Sort in text editor. In fact to do this in a simple manner it may be only one option to shos or hide Codemaid items in the solution explorer context menu.

:bulb: *Another easy way may be to group Codemaid items in the solution explorer root context menu in a single menu as in the Text editor.*

:memo: I would certainly create issues in the future for these suggestions if not exists but here they are quickly:
* Text Editor : Cleanup selected text.
* Text Editor : Remove blank lines in selection (no rules, just remove all blank lines).
* Spade : Rename (open the refactor renaming VS box).

**Comment by codecadwallader** [comment]
@Ordisoftware thanks for the comments and suggestions!  The PR for making features switchable has been accepted and is in testing.  If you want to help test an early version of it you could get it from our CI server here: http://vsixgallery.com/extension/4c82e17d-927e-42d2-8460-b473ac7df316/

Creating separate issues for your suggestions (and performing a quick search to try to avoid duplicates) would be great.

**Comment by heku** [comment]
@codecadwallader Are you able to open the CI server link? I open it but see nothing. I'm wondering if it's my network problem.

**Comment by codecadwallader** [comment]
@heku Yes it is working for me.  I have been pushing a couple minor changes this morning, perhaps just try refreshing?

**Comment by Ordisoftware** [comment]
@codecadwallader Installed CodeMaid v10.4.85.vsix and it works at first sight, thank you. I disabled Cleanup all code and cleanup selected code while able to use all others from main menu or text editor submenu. I'll do check and add suggestions but I saw the refactor link from code digger is already submitted. All the best.

**Comment by codecadwallader** [comment]
@Ordisoftware Wonderful, thank you for testing!

**Comment by codecadwallader** [comment]
@heku That makes sense, although it is disappointing.  At least you have access to the code and can compile it yourself so that is a good workaround.
